clients/pubsub-common/src/test/java/org/eclipse/hono/client/pubsub/AbstractPubSubBasedMessageSenderTest.java

```diff
@@ -0,0 +1,196 @@
+
+                });
+
+        assertThrows(NullPointerException.class,
+                () -> new AbstractPubSubBasedMessageSender(factory, null, PROJECT_ID, tracer) {
+
+                });
+
+        assertThrows(NullPointerException.class,
+                () -> new AbstractPubSubBasedMessageSender(factory, TOPIC, null, tracer) {
+
+                });
+
+        assertThrows(NullPointerException.class,
+                () -> new AbstractPubSubBasedMessageSender(factory, TOPIC, PROJECT_ID, null) {
+
+                });
+    }
+
+    /**
+     * Verifies that the send method returns a failed future wrapped in a {@link ServerErrorException}.
+     */
+    @Test
+    public void testSendFailedWhenLifecycleStatusIsNotStarted() {
+        sender.stop();
+
+        final var result = sender.sendAndWaitForOutcome(TOPIC, TENANT_ID, DEVICE_ID, null, Map.of(), NoopSpan.INSTANCE);
+
+        assertThat(result.failed()).isTrue();
+        result.onFailure(t -> assertThat(t.getClass()).isEqualTo(ServerErrorException.class));
+    }
+
+    /**
+     * Verifies that the send method throws a {@link ServerErrorException}.
+     */
+    @Test
+    public void testSendFailed() {
+        final PubSubPublisherClient client = mock(PubSubPublisherClient.class);
+        when(client.publish(Mockito.any())).thenReturn(
+                Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_CONFLICT, "client is null")));
+        when(factory.getOrCreatePublisher(TOPIC, PROJECT_ID, TENANT_ID)).thenReturn(client);
+
+        try (MockedStatic<PubsubMessage> pubSubMessage = Mockito.mockStatic(PubsubMessage.class)) {
+            final PubsubMessage.Builder builder = mock(PubsubMessage.Builder.class);
+            when(builder.putAllAttributes(Mockito.any())).thenReturn(builder);
+            when(builder.setOrderingKey(DEVICE_ID)).thenReturn(builder);
+            when(builder.setData(Mockito.any())).thenReturn(builder);
+
+            final PubsubMessage message = mock(PubsubMessage.class);
+            when(builder.build()).thenReturn(message);
+
+            pubSubMessage.when(PubsubMessage::newBuilder).thenReturn(builder);
+
+            assertThrows(NullPointerException.class,
+                    () -> sender.sendAndWaitForOutcome(TOPIC, TENANT_ID, DEVICE_ID, null, Map.of(), NoopSpan.INSTANCE));
+
+            assertThrows(ServerErrorException.class,
+                    () -> sender.sendAndWaitForOutcome(TOPIC, TENANT_ID, DEVICE_ID, Buffer.buffer("test payload"),
+                            Map.of(), NoopSpan.INSTANCE));
+        }
+
+    }
+
+    /**
+     * Verifies that the send method succeeded.
+     */
+    @Test
+    public void testSendSucceeded() {
+        final Map<String, Object> properties = getProperties();
+        final Map<String, String> attributes = getAttributes();
+
+        final PubSubPublisherClient client = mock(PubSubPublisherClient.class);
```

Comment:
instead of having to mock a final class, wouldn't it be better to have *PubSubPublisherFactory.getOrCreatePublisher()* return a *PublisherInterface* instead of a *PubSubPublisherClient* and let *PubSubPublisherClient* implement *PublisherInterface*? That would make testing a lot easier, I assume ...

Reply:
Sure, I added a new interface PubSubPublisherClient & renamed the implementation to PubSubPublisherClientImpl